CIPs/0017.md

```diff
@@ -0,0 +1,96 @@
+
+When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propagate the proof throughout the network.
+
+This `AddProof` function will then utilize the `verificationKey` corresponding to the `versionNumber` and the corresponding validator sets for these epochs to verify the proof. If successful, it will then store the proof in a new database `plumoproofdata`, keyed by it's combination of first and last epoch. 
+
+To propagate these proofs throughout the network, we also propose to extend the `eth` protocol with three new message types: 
+```
+NewPlumoProof (0x019)
+[[firstEpoch, lastEpoch, versionNumber], ...]
+Specify one or more new plumo proofs which have appeared on the network.
+This is intended to be honored with a proceeding GetPlumoProofsMsg to request the proof.
+
+GetPlumoProofsMsg (0x1a)
+[[firstEpoch, lastEpoch, versionNumber], ...]
+Sent from node A to node B with the set of proofs that node A currently has, requires node B to return a PlumoProofsMsg.
+Node B's reply must contain any proofs that node B has that node A did not send.
+
+PlumoProofMsg (0x1b)
+[plumoProof_0, plumoProof_1, ...]
+Reply to GetPlumoProofMsg. The items in the list are plumo proof structs, {proof: []byte, firstEpoch: uint, lastEpoch: uint, versionNumber: uint}.
+This may validly contain no plumo proofs if the GetPlumoProofMsg included all proofs this node is aware of.
+```
+
+The `les` protocol must then also be extended via at least two messages:
+```
+GetCertifiedEpochsValidatorSet - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.
+CertifiedEpochsValidatorSet - Receives a mix of header(s) and proof(s) from the server.
+```
+
+The `les` client must then be modified to send a `GetCertifiedEpochsValidatorSet` message during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`.
+
+Upon receiving the `CertifiedEpochsValidatorSet` response message, the light client must then verify headers and proofs provided. Proofs will be verified via a call into `celo-bls-go`'s `verify` fn.
+
+The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.
+
+### Upgradeability
+In order to support an upgradeable snark circuit, we utilize a `versionNumber` parameter with every proof. This corresponds to a version of the snark circuit and thus a corresponding verification key. These version numbers will be stored on chain, within the `blockchain_parameters` smart contract, with the corresponding hash of the verification key. The clients are expected to store all applicable verification keys.
```

Comment:
Doesn't storing the version numbers on-chain create a chicken-and-egg problem in that the user of a plumo proof is not synced to the chain yet? It may be necessary to ship verification keys in the client code instead.

Reply:
This is a great point - I was thinking we could ship the VKs in the client code, but then have the clients verify each key against the stored reference hash on chain, but of course if you aren't synced you won't see that hash.

I agree, I thought through some plans to allow for this, but if we want to use chain state to verify keys then I believe you need to set boundaries for verification keys, i.e. you can only use version n to prove transitions on epochs greater than X (since you publish the hash to the chain in epoch X, can't access it before). I think this is worse than just shipping all VKs with the client like you mentioned, since then we can upgrade our circuit and use it for any epochs, allowing us to replace a potentially inferior proof scheme.